pkg/solana/chainwriter/lookups.go

```diff
@@ -0,0 +1,362 @@
+		}
+
+		decoded, err := decodeBorshIntoType(accountInfo.GetBinary(), pda.InternalField.Type)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding Borsh data dynamically: %w", err)
+		}
+
+		value, err := GetValuesAtLocation(decoded, pda.InternalField.Location)
+		if err != nil {
+			return nil, fmt.Errorf("error getting value at location: %w", err)
+		}
+		if len(value) > 1 {
+			return nil, fmt.Errorf("multiple values found at location: %s", pda.InternalField.Location)
+		}
+
+		result = append(result, &solana.AccountMeta{
+			PublicKey:  solana.PublicKeyFromBytes(value[0]),
+			IsSigner:   accountMeta.IsSigner,
+			IsWritable: accountMeta.IsWritable,
+		})
+	}
+	return result, nil
+}
+
+func decodeBorshIntoType(data []byte, typ reflect.Type) (interface{}, error) {
+	// Ensure the type is a struct
+	if typ.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("provided type is not a struct: %s", typ.Kind())
+	}
+
+	// Create a new instance of the type
+	instance := reflect.New(typ).Interface()
+
+	// Decode using Borsh
+	err := ag_binary.NewBorshDecoder(data).Decode(instance)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding Borsh data: %w", err)
+	}
+
+	// Return the underlying value (not a pointer)
+	return reflect.ValueOf(instance).Elem().Interface(), nil
+}
+
+// getSeedBytes extracts the seeds for the PDALookups.
+// It handles both AddressSeeds (which are public keys) and ValueSeeds (which are byte arrays from input args).
+func getSeedBytes(ctx context.Context, lookup PDALookups, args any, derivedTableMap map[string]map[string][]*solana.AccountMeta, reader client.Reader) ([][]byte, error) {
+	var seedBytes [][]byte
+	maxSeedLength := 32
+
+	for _, seed := range lookup.Seeds {
+		if lookupSeed, ok := seed.(AccountLookup); ok {
+			// Get value from a location (This doens't have to be an address, it can be any value)
+			bytes, err := GetValuesAtLocation(args, lookupSeed.Location)
+			if err != nil {
+				return nil, fmt.Errorf("error getting address seed: %w", err)
+			}
+			// validate seed length
+			for _, b := range bytes {
+				if len(b) > maxSeedLength {
+					return nil, fmt.Errorf("seed byte array exceeds maximum length of %d: got %d bytes", maxSeedLength, len(b))
+				}
+				seedBytes = append(seedBytes, b)
+			}
+		} else {
+			// Get address seeds from the lookup
+			seedAddresses, err := GetAddresses(ctx, args, []Lookup{seed}, derivedTableMap, reader)
+			if err != nil {
+				return nil, fmt.Errorf("error getting address seed: %w", err)
+			}
+
+			// Add each address seed as bytes
+			for _, address := range seedAddresses {
+				seedBytes = append(seedBytes, address.PublicKey.Bytes())
+			}
+		}
+	}
+
+	return seedBytes, nil
+}
+
+// generatePDAs generates program-derived addresses (PDAs) from public keys and seeds.
+func generatePDAs(publicKeys []*solana.AccountMeta, seeds [][]byte, lookup PDALookups) ([]*solana.AccountMeta, error) {
+	if len(seeds) >= 16 {
```

Comment:
Since 16 seeds is allowed, think this check should just be >
```suggestion
	if len(seeds) > 16 {
```